readme.md
# Graphene for Lua
![shield_license]
![shield_build]
![shield_tests]

![shield_release_version]
![shield_prerelease_version]
![shield_dev_version]

Graphene is a library intended to make creation of large Lua libraries easier by providing a common structure to handle library includes in a sane way. It provides namespacing semantics similar to .NET that map directly to the filesystem and provides a large amount of platform feature detection.

To create a library intended for use with Graphene, rename `graphene.lua` to `init.lua` and put it in your library's root folder. Then, when the folder containing your library is loaded, a Graphene namespace will be returned that handles loading the rest of your library. What used to be `init.lua` in a library should then become `_.lua`, which is automatically loaded by Graphene to initialize directories.

See [Rationale](docs/rationale.md) for an argument for, and potentially against Graphene.

## Support
Graphene presently runs under the following platforms natively:
- Lua 5.1
- Lua 5.2/5.3 with debug library
- LuaJIT 2.0+
- LÖVE 0.9.0+

LuaFileSystem is required for advanced filesystem features like directory listing and directory importing. On 5.2 and 5.3, the debug library is required for imports to function due to changes in the environment system.

Any platforms not officially supported can use the included virtual filesystem, which can also be used to pack a library up into a single file. This functionality is deprecated, however, and will be included instead as an extra in a future version of Graphene.

## Feature Detection
Graphene feature detects for the following features and exposes them through `Graphene.Support`:
- Lua 5.1, 5.2, 5.3+ (`lua51`, `lua52`, `lua53`)
- LuaJIT 2.0+ (`luajit` in 1.1.9+)
- LÖVE 0.9.0+ (`love`)
- Debug library (`debug`)
- IO library (`io`)
- OS library (`os`)
- LuaFileSystem (`lfs`)
- LuaJIT FFI / LuaFFI (`ffi`)
- xpcall with passthrough arguments (`xpcallargs`)

## Simple Usage

Libraries using Graphene will have a source similar to:

library/one.lua:
```lua
local library = (...)
local two = library.two

local one = {
	name = "one"
}

return one
```

And then code using the library would look like this:

```lua
local library = require("library")
print(library.one.name) --> one
```

[shield_license]: https://img.shields.io/badge/license-zlib/libpng-333333.svg?style=flat-square
[shield_build]: https://img.shields.io/badge/build-unknown-lightgrey.svg?style=flat-square
[shield_tests]: https://img.shields.io/badge/tests-0/0-lightgrey.svg?style=flat-square
[shield_release_version]: https://img.shields.io/badge/release-1.1.11-brightgreen.svg?style=flat-square
[shield_prerelease_version]: https://img.shields.io/badge/prerelease-none-lightgrey.svg?style=flat-square
[shield_dev_version]: https://img.shields.io/badge/development-2.0.0-orange.svg?style=flat-square

changelog.md
# Change Log

![1.1.11](https://img.shields.io/badge/1.1.11-latest-brightgreen.svg?style=flat-square)
- Fixed directory objects being polluted by Graphene fields
- Added `G:MakeLightImportable` for making a lighter-footprint importable

![1.1.10](https://img.shields.io/badge/1.1.10-unsupported-red.svg?style=flat-square)
- Fixed issue with nested submodules due to submodule path specificity

![1.1.9](https://img.shields.io/badge/1.1.9-unsupported-red.svg?style=flat-square)
- Added `luajit`, `ffi` features
- Fixed metadata to have weak keys
- Deprecated `jit` feature
- Deprecated internal VFS
- Deprecated 'Lib' switch

![1.1.8](https://img.shields.io/badge/1.1.8-unsupported-red.svg?style=flat-square)
- Fixed `MakeImportable` causing issues with init (`_`) files

![1.1.7](https://img.shields.io/badge/1.1.7-unsupported-red.svg?style=flat-square)
- Fixed extraneous debug print

![1.1.6](https://img.shields.io/badge/1.1.6-unsupported-red.svg?style=flat-square)
- Fixed LOVE file detection

![1.1.5](https://img.shields.io/badge/1.1.5-unsupported-red.svg?style=flat-square)
- Fixed loading on Linux
- Fixed FullyLoad loading from `..` on LOVE

![1.1.4](https://img.shields.io/badge/1.1.4-unsupported-red.svg?style=flat-square)
- Fixed Lua 5.1 compatibility

![1.1.3](https://img.shields.io/badge/1.1.3-unsupported-red.svg?style=flat-square)
- Fix :FullyLoad when called on the root element

![1.1.2](https://img.shields.io/badge/1.1.2-unsupported-red.svg?style=flat-square)
- Fix file loader for Lua 5.1

![1.1.1](https://img.shields.io/badge/1.1.1-unsupported-red.svg?style=flat-square)
- Add LFS Toggle Flag
- Fix LOVE printing when using LuaJIT

![1.1.0](https://img.shields.io/badge/1.1.0-unsupported-red.svg?style=flat-square)
- Numerous filesystem fixes for LOVE
- Modules are now passed a `metadata` field that gives runtime information to the module.
- FS providers now fill an `FSID` field that gives the ID of the filesystem abstraction.
- FS providers now fill a `Loader` field determining what function should load the file. Defaults to a compatibility shim for `load`.
- FS providers now internally use a common method for pathing, making way for frontend support.
- FS providers no longer have to provide `IsFile` or `IsDirectory` methods, but these have been deprecated since 1.0.0.
- Added `G:SetLoadCallback` and `G:GetLoadCallback` for setting notification of loaded modules.
- Added `G:SetErrorCallback` and `G:SetErrorCallback` for setting notification of errors in modules.
- Added `G:SetSafeMode` for enabling automatic error catching on loaded modules.
- Added `G.Metadata` for storing/retrieving object metadata.
- Added `G.Directory` reference to local directory object.
- Added `Directory:Import`, `Directory:ImportAs`, `Directory:ImportAll`, and `G:MakeImportable` for importing loaded modules.
- `G:Get` now returns a second item, the metadata of the loaded module, if it exists.
- Removed `G.Config.FileExtensions`, replaced with `G.Config.Loaders`: it was undocumented anyways.
- Added support for custom Lua frontends, like MoonScript, with `G.Config.Loaders`.
- Improved LuaFileSystem fallbacks for Windows (notably for checking directory existence)
- Support modules now use public-style naming as they were supposed to in 1.0.
- Deprecated methods, these will be removed in 2.0:
	- `Directory:GetGrapheneCore`, use `Directory:GetGraphene` instead.
	- `G:ClearRebases`, use `G:ClearSubmodules` instead.
- Directories now have an "IsDirectory" member to mark them as Graphene directories.
- Internal renaming: this should not affect existing modules
- Improved Lua 5.2 compatiblity.

![1.0.1](https://img.shields.io/badge/1.0.1-unsupported-red.svg?style=flat-square)
- Fixed version check with LOVE 0.9.0

![1.0.0](https://img.shields.io/badge/1.0.0-unsupported-red.svg?style=flat-square)
- Renamed library to Graphene.
- `GetNamespaceCore` renamed to `GetGrapheneCore`.
- Recommended variable to reference the graphene core is now G.
- Based on Namespace 0.4.0-beta.
- New `gengraphene.lua` generator for slimming down the Graphene core.
- Now supports multiple file extensions; see `G.Config.FileExtensions`.
- API methods are now type checked to some degree.
- Rebasing rules have turned into Submodules, and `G:AddRebase` has been renamed to `G:AddSubmodule`.
- Added GenGraphine, a command line tool to generate smaller versions of Graphene.

test/_.lua
return {
	name = "test"
}

run-tests.lua
--[[
	Graphene: Run all tests

	Runs all tests in order.
	Overrides print to make it prettier for the tests.
]]

local real_print = print

print = function(...)
	local args = {}

	for i = 1, select("#", ...) do
		table.insert(args, "\t" .. tostring(select(i, ...)):gsub("\r?\n", "\n\t"))
	end

	return real_print(unpack(args))
end

real_print("Graphene: Running all tests...")

require("tests.vfs")
require("tests.loader")

real_print("Graphene: All tests passed.")

print = real_print

tests/vfs.lua
--[[
	Graphene VFS Test

	Makes sure the internal VFS functions correctly.
]]

print("Graphene: Starting VFS Test...")

local Graphene = require("graphene")
local G = Graphene:GetGrapheneCore()
local vfs = G.FS:GetProvider("vfs")

vfs:Clear()

vfs:AddFile("ABC", "ABC")

vfs:AddDirectory("Test")
vfs:AddFile("Test.ABC", "Test.ABC")
vfs:AddFile("Test.XYZ", "Test.XYZ")

vfs:AddFile("Hello.World", "Hello.World")

local ABC = vfs:GetFile("ABC")
local Test = vfs:GetDirectory("Test")
local Test_ABC = vfs:GetFile("Test.ABC")
local Test_XYZ = vfs:GetFile("Test.XYZ")

local Hello = vfs:GetDirectory("Hello")
local Hello_World = vfs:GetFile("Hello.World")

assert(ABC, "Could not retrieve explicitly initialized file 'ABC'.")
assert(ABC:Read() == "ABC", "File 'ABC' did not preserve contents.")

assert(Test, "Could not retrieve explicitly initialized directory 'Test'.")
assert(Test_ABC and Test_XYZ, "Could not retrieve one or more explicitly initialized files ('Test.ABC', 'Test.XYZ').")
assert(Test_ABC:Read() == "Test.ABC", "File 'Test.ABC' did not preserve contents.")
assert(Test_XYZ:Read() == "Test.XYZ", "File 'Test.XYZ' did not preserve contents.")

assert(Hello, "Could not retrieve implicitly initialized directory 'Hello'.")
assert(Hello_World, "Could not retrieve explicitly initialized file 'Hello.World'.")

vfs:Clear()

print("Graphene: VFS test passed.")

namespace.lua
--[[
	Lua Namespace 0.4.0

	Copyright (c) 2014 Lucien Greathouse (LPGhatguy)

	This software is provided 'as-is', without any express or implied warranty.
	In no event will the authors be held liable for any damages arising from the
	use of this software.

	Permission is granted to anyone to use this software for any purpose, including
	commercial applications, and to alter it and redistribute it freely, subject to
	the following restrictions:

	1. The origin of this software must not be misrepresented; you must not claim
	that you wrote the original software. If you use this software in a product, an
	acknowledgment in the product documentation would be appreciated but is not required.

	2. Altered source versions must be plainly marked as such, and must not be misrepresented
	as being the original software.

	3. This notice may not be removed or altered from any source distribution.
]]

-- Current namespace version
local n_version = {0, 4, 0, "beta"}
local n_versionstring = ("%s.%s.%s-%s"):format((unpack or table.unpack)(n_version))

-- Determine Lua capabilities and library support
local support = {}

--[[
	string Support:Report()

	Generates a stringified report of supported features and extensions.
]]
function support:Report()
	local features = {}

	for feature, enabled in pairs(self) do
		if (type(enabled) == "table" and enabled.report) then
			table.insert(features, enabled:report())
		elseif (type(enabled) ~= "function") then
			table.insert(features, feature)
		end
	end

	return table.concat(features, ", ")
end

-- Contains our actual core
local N = {
	_loaded = {}, -- Dictionary of loaded modules for caching
	_rebasing = {}, -- Contains rebasing information
	Support = support, -- Table for fast support lookups

	Version = n_version, -- Version table for programmatic comparisons
	VersionString = n_versionstring, -- Version string for user-facing reporting

	-- Filesystem abstraction
	FS = {
		Providers = {}
	},

	-- Configuration
	Config = {
		Lib = true
	}
}

-- Do we have LFS?
local ok, lfs = pcall(require, "lfs")
if (not ok) then
	lfs = nil
end

-- Do we have HATE?
local ok, hate = pcall(require, "hate")
if (not ok) then
	hate = nil
end

-- What Lua are we running under?
-- For our purposes, 5.3 is a superset of 5.2.
if (table.unpack) then
	support.lua52 = true

	if (table.move) then
		support.lua53 = true
	end
else
	support.lua51 = true
end

-- LuaJIT 2.0+
-- Also check for OS in this path.
if (jit) then
	support.jit = true

	if (jit.os == "Windows") then
		support.windows = true
	else
		support.nix = true
	end
else
	local win = package.config:sub(1, 1) == "\\"

	if (win) then
		support.windows = true
	else
		support.nix = true
	end
end

-- LuaFileSystem
if (lfs) then
	support.lfs = true
end

-- Lua debug library
if (debug) then
	support.debug = true
end

-- Lua os library
if (os) then
	support.os = true
end

-- Lua io library
if (io) then
	support.io = true
end

-- Is hate available?
if (hate) then
	support.hate = true
end

-- Are we running in LOVE?
if (love) then
	support.love = {}

	-- return a nice report of the current LOVE version
	function support.love:report()
		return ("love %d.%d.%d (%s)"):format(unpack(self.version))
	end

	if (love.getVersion) then
		-- LOVE 0.9.1+
		support.love.Version = {love.getVersion()}
	else
		-- LOVE 0.9.0 and older; may be supported
		local major = love._version_minor
		local minor = love._version_minor

		-- What *IS* LOVE?
		if (major ~= 0) then
			support.love = false
		end

		if (minor == 9) then
			-- Definitely 0.9.0
			support.love.Version = {0, 9, 0, "Baby Inspector"}
		else
			-- 0.8.0 and older; definitely not supported
			support.love = false
		end
	end
end

-- How about ROBLOX?
if (game and workspace and Instance) then
	support.roblox = true
end

-- Cross-version shims
local unpack = unpack or table.unpack

--[[
	loaded load_with_env(string source, table environment)
		source: The source code to compile into a function.
		environment: The environment to load the function into.

	Loads a function with a given environment.
	Essentially backports Lua 5.2's load function to LuaJIT and Lua 5.1.
]]
local load_with_env

if (support.lua51) then
	function load_with_env(source, from, environment)
		environment = environment or getfenv()
		local chunk, err = loadstring(source, from)

		if (not chunk) then
			return chunk, err
		end

		setfenv(chunk, environment)

		return chunk
	end
elseif (support.lua52) then
	load_with_env = function(source, from, environment)
		return load(source, from, nil, environment)
	end
end

-- Provider is_directory and is_file fallbacks for systems without LFS.
local is_directory
local is_file

if (support.lfs) then
	function is_file(path)
		return (lfs.attributes(path, "mode") == "file")
	end

	function is_directory(path)
		return (lfs.attributes(path, "mode") == "directory")
	end
else
	-- Reduced file functionality without LFS
	function is_file(path)
		local handle = io.open(path, "r")

		if (handle) then
			handle:close()

			return true
		end

		return false
	end

	if (support.windows) then
		function is_directory(path)
			return (os.execute(("cd %q 2>nul"):format(path)) == 0)
		end
	else
		function is_directory(path)
			return (os.execute(("stat %q"):format(path)) == 0)
		end
	end
end

-- Find out a path for the directory above namespace
local n_root
local n_file = support.debug and debug.getinfo(1, "S").source:match("@(.+)$")

if (n_file) then
	-- Normalize slashes; this is okay for Windows
	n_root = n_file:gsub("\\", "/"):match("^(.+)/.-$") or "./"
else
	print("Could not locate lua-namespace source file; is debug info stripped?")
	print("This code path is untested.")
	n_root = (...):match("(.+)%..-$")
end

-- Utility Methods

--[[
	module_to_file(string source, bool is_directory=false)
		source: The module path to parse
		is_directory: Whether the output should be a file or directory.

	Takes a module path (a.b.c) and turns it into a somewhat well-formed file path.
	If is_directory is true, the output will look like:
		a/b/c
	Otherwise:
		a/b/c.lua
]]
local function module_to_file(source, is_directory)
	return (source:gsub("%.", "/") .. (is_directory and "" or ".lua"))
end

--[[
	(string module, bool is_directory) file_to_module(string source)
		source: The file path to be turned into a module path.

	Takes a file path (a/b/c or a/b/c.ext) and turns it into a well-formed module path.
	Also returns whether the file is most likely a directory object or not.
]]
local function file_to_module(source)
	local fsource = source:gsub("%..-$", "")
	local is_file = (fsource ~= source)
	return (fsource:gsub("[/\\]+", "."):gsub("^%.*", ""):gsub("%.*$", "")), is_file
end

--[[
	string module_join(string first, string second)
		first: The first part of the path.
		second: The second part of the path.

	Joins two module names with a period and removes any extraneous periods.
]]
local function module_join(first, second)
	return ((first .. "." .. second):gsub("%.%.+", "."):gsub("^%.+", ""):gsub("%.+$", ""))
end

--[[
	string path_join(string first, string second)
		first: The first part of the path.
		second: The second part of the path.

	Joins two path names with a period and removes any extraneous slashes.
]]
local function path_join(first, second)
	return ((first .. "/" .. second):gsub("//+", "/"):gsub("/+$", ""))
end

--[[
	table dictionary_shallow_copy(table from, table to)
		from: The table to source data from.
		to: The table to copy data into.

	Performs a shallow copy from one table to another.
]]
local function dictionary_shallow_copy(from, to)
	to = to or {}

	for key, value in pairs(from) do
		to[key] = value
	end

	return to
end

-- Filesystem Abstractions

--[[
	File? FS:GetFile(string path)
		path: The file to find a provider for.

	Returns the file from whatever filesystem provider it's located on.
]]
function N.FS:GetFile(path)
	for i, provider in ipairs(self.Providers) do
		if (provider.GetFile) then
			local file = provider:GetFile(path)

			if (file) then
				return file
			end
		end
	end

	return nil
end

--[[
	directory? FS:GetDirectory(string path)
		path: The directory to find a provider for.

	Returns the directory from whatever filesystem provider it's located on.
]]
function N.FS:GetDirectory(path)
	for i, provider in ipairs(self.Providers) do
		if (provider.GetDirectory) then
			local directory = provider:GetDirectory(path)

			if (directory) then
				return directory
			end
		end
	end

	return nil
end

--[[
	provider? FS:GetProvider(string id)
		id: The ID of the FS provider to search for.

	Returns the provider with the given ID if it exists.
]]
function N.FS:GetProvider(id)
	for i, provider in ipairs(self.Providers) do
		if (provider.ID == id) then
			return provider
		end
	end

	return nil
end

--[[
	Filesystem provider schema:

	Provider.ID
		An ID to look up the provider with.

	Provider.Name
		A friendly name to describe the provider.

	Provider.Path (LOVE and vanilla IO only)
		Similar to a system PATH, where the provider looks for files.

	bool Provider:IsFile(string path)
			path: The module path to check.

	Returns whether the specified path exists on this filesystem provider.

	File Provider:GetFile(string path)
		path: The module path to check.

		Returns a file object corresponding to the given file on this filesystem.

	bool Provider:IsDirectory(string path)
		path: The module path to check.

		Returns whether the specified path exists on this filesystem provider.

	Directory Provider:GetDirectory(string path)
		path: The module path to check.

		Returns a directory object corresponding to the given directory on this filesystem.

	string File:Read()
		contents: The complete contents of the file.

		Reads the entire file into a string and returns it.

	void File:Close()
	
		Closes the file, allowing it to be reused by the system.

	string[] Directory:List()
		files: The files and folders contained in this directory.

		Returns a list of files contained in the directory.

	void Directory:Close()
	
		Closes the directory, allowing it to be reused by the system.
]]

-- LOVE filesystem provider
if (support.love) then
	local love_fs = {
		ID = "love",
		Name = "LOVE Filesystem",
		Path = {"", n_root}
	}

	table.insert(N.FS.providers, love_fs)

	local file_buffer = {}
	local directory_buffer = {}

	local function file_close(self)
		table.insert(file_buffer, self)
	end

	local function file_read(self)
		return love.filesystem.read(self.filepath)
	end

	local function directory_close(self)
		table.insert(directory_buffer, self)
	end

	local function directory_list(self)
		local items = love.filesystem.getDirectoryItems(self.filepath)

		for i = 1, #items do
			items[i] = module_join(self.path, file_to_module(items[i]))
		end

		return items
	end

	function love_fs:GetFile(path, filepath)
		filepath = filepath or module_to_file(path)

		for i, base in ipairs(self.Path) do
			local fullpath = path_join(base, filepath)

			if (self:IsFile(path, fullpath)) then
				local file = file_buffer[#file_buffer]
				file_buffer[#file_buffer] = nil

				if (file) then
					file.Path = path
					file.FilePath = fullpath

					return file
				else
					return {
						Close = file_close,
						Read = file_read,
						Path = path,
						FilePath = fullpath
					}
				end
			end
		end
	end

	function love_fs:IsFile(path, filepath)
		filepath = filepath or module_to_file(path)

		return love.filesystem.isFile(filepath)
	end

	function love_fs:GetDirectory(path, filepath)
		filepath = filepath or module_to_file(path, true)

		for i, base in ipairs(self.Path) do
			local fullpath = path_join(base, filepath)

			if (self:IsDirectory(path, fullpath)) then
				local directory = directory_buffer[#directory_buffer]
				directory_buffer[#directory_buffer] = nil

				if (directory) then
					directory.Path = path
					directory.FilePath = fullpath

					return directory
				else
					return {
						Close = directory_close,
						List = directory_list,
						Path = path,
						FilePath = fullpath
					}
				end
			end
		end
	end

	function love_fs:IsDirectory(path, filepath)
		filepath = filepath or module_to_file(path, true)

		return love.filesystem.isDirectory(filepath)
	end
end

-- Only support the full filesystem if we have IO
-- No FullyLoad method without LFS
-- FS provider to read from the actual filesystem
if (support.io) then
	local full_fs = {
		ID = "io",
		Name = "Full Filesystem",
		Path = {"", n_root}
	}

	table.insert(N.FS.Providers, full_fs)

	local file_buffer = {}
	local directory_buffer = {}

	-- File:Close() method
	local function file_close(self)
		table.insert(file_buffer, self)
	end

	-- File:Read() method
	local function file_read(self)
		local handle, err = io.open(self.FilePath, "r")

		if (handle) then
			local body = handle:read("*a")
			handle:close()

			return body
		else
			return nil, err
		end
	end

	-- Directory:Close() method
	local function directory_close(self)
		table.insert(directory_buffer, self)
	end

	-- Directory:List() method
	local function directory_list(self)
		if (not support.lfs) then
			error("Cannot list directory without LFS!", 2)
		end

		local paths = {}

		for name in lfs.dir(self.FilePath) do
			if (name ~= "." and name ~= "..") then
				table.insert(paths, (file_to_module(name)))
			end
		end

		return paths
	end

	function full_fs:GetFile(path, filepath)
		filepath = filepath or module_to_file(path)

		for i, base in ipairs(self.Path) do
			local fullpath = path_join(base, filepath)

			if (self:IsFile(path, fullpath)) then
				local file = file_buffer[#file_buffer]
				file_buffer[#file_buffer] = nil

				if (file) then
					file.Path = path
					file.FilePath = filepath

					return file
				else
					return {
						Close = file_close,
						Read = file_read,
						Path = path,
						FilePath = fullpath
					}
				end

				break
			end
		end
	end

	-- Is this a file?
	function full_fs:IsFile(path, filepath)
		filepath = filepath or module_to_file(path)

		return is_file(filepath)
	end

	-- Returns a directory object
	function full_fs:GetDirectory(path, filepath)
		filepath = filepath or module_to_file(path, true)

		for i, base in ipairs(self.Path) do
			local fullpath = path_join(base, filepath)

			if (self:IsDirectory(path, fullpath)) then
				local directory = directory_buffer[#directory_buffer]
				directory_buffer[#directory_buffer] = nil

				if (directory) then
					directory.Path = path
					directory.FilePath = fullpath

					return directory
				else
					return {
						Close = directory_close,
						List = directory_list,
						Path = path,
						FilePath = fullpath
					}
				end

				break
			end
		end
	end

	-- Is this a directory?
	function full_fs:IsDirectory(path, filepath)
		filepath = filepath or module_to_file(path, true)

		return is_directory(filepath)
	end
end

-- ROBLOX "filesystem" provider
-- TODO
if (support.roblox) then
end

-- Virtual Filesystem for namespace
-- Used when packing for platforms that don't have real filesystem access
do
	local vfs = {
		ID = "vfs",
		Name = "Virtual Filesystem",
		Enabled = false,

		Nodes = {},
		Directory = true
	}

	table.insert(N.FS.Providers, vfs)

	-- File:Read() method
	local function file_read(self)
		return self._contents
	end

	-- File:Close() method
	-- a stub, since this doesn't apply to a VFS
	local function file_close()
	end

	-- Directory:List() method
	local function directory_list(self)
		local list = {}

		for name in pairs(self._nodes) do
			table.insert(list, name)
		end

		return list
	end

	-- Directory:Close() method
	-- a stub, since this doesn't apply to a VFS
	local function directory_close()
	end

	-- Starts at a root node and navigates according to a module name
	-- Add auto_dir to automatically create directories
	-- If the path does not exist, or cannot be reached due to an invalid node,
	-- the function will return nil, the furthest location reached, and a list of node names navigated.
	function vfs:Navigate(path, auto_dir)
		local location = self
		local nodes = {}

		for node in path:gmatch("[^%.]+") do
			if (not location.Nodes) then
				return nil, location, nodes
			end

			if (location.Nodes[node]) then
				location = location.Nodes[node]
				table.insert(nodes, node)
			elseif (auto_dir) then
				location = vfs:AddDirectory(table.concat(nodes, "."))
				table.insert(nodes, node)
			else
				return nil, location, nodes
			end
		end

		return location
	end

	-- Performs string parsing and navigates to the parent node of a given path
	function vfs:LeafedNavigate(path, auto_dir)
		local leafless, leaf = path:match("^(.-)%.([^%.]+)$")

		if (leafless) then
			local parent = self:Navigate(leafless, auto_dir)

			-- Couldn't get there! Ouch!
			if (not parent) then
				return nil, ("Could not navigate to parent node '%s': invalid path"):format(leafless)
			end

			if (not parent.Directory) then
				return nil, ("Could not create node in node '%s': not a directory"):format(leafless)
			end

			return parent, leafless, leaf
		else
			leafless = ""
			leaf = path

			return self, leafless, leaf
		end
	end

	function vfs:GetFile(path)
		if (not self.Enabled) then
			return false
		end

		local object = self:Navigate(path)

		if (object and object.File) then
			return {
				_contents = object.Contents,
				Read = file_read,
				Close = file_close,
				Path = path
			}
		end
	end

	function vfs:IsFile(path)
		if (not self.Enabled) then
			return false
		end

		local object = self:Navigate(path)

		return (object and object.File)
	end

	function vfs:AddFile(path, contents)
		self.Enabled = true
		local parent, leafless, leaf = self:LeafedNavigate(path, true)

		-- leafless contains error state if parent is nil
		if (not parent) then
			return nil, leafless
		end

		local node = {
			File = true,
			Contents = contents
		}

		parent.Nodes[leaf] = node

		return node
	end

	function vfs:GetDirectory(path)
		if (not self.Enabled) then
			return false
		end

		local object = self:Navigate(path)

		if (object and object.Directory) then
			return {
				_nodes = object.Nodes,
				List = directory_list,
				Close = directory_close,
				Path = path
			}
		end
	end

	function vfs:IsDirectory(path)
		if (not self.Enabled) then
			return false
		end

		local object = self:Navigate(path)

		return (object and object.Directory)
	end

	function vfs:AddDirectory(path)
		self.Enabled = true
		local parent, leafless, leaf = self:LeafedNavigate(path, true)

		-- leafless contains error state if parent is nil
		if (not parent) then
			return nil, leafless
		end

		local node = {
			Directory = true,
			Nodes = {}
		}

		parent.Nodes[leaf] = node

		return node
	end
end

local directory_interface = {}

--[[
	N Directory:GetNamespaceCore()

	Returns the namespace core, defined as N in this file.
	Not affected by any basing, rebasing, or such.
]]
function directory_interface:GetNamespaceCore()
	return N
end

--[[
	void Directory:FullyLoad()

	Recursively loads all members of the directory.
]]
function directory_interface:FullyLoad()
	local list = self._directory:List()

	for i, member in ipairs(list) do
		local object = self[member]

		-- Make sure we have an object that isn't this one (necessary because of _.lua).
		-- Also make sure that it's got a FullyLoad method, which makes it either a directory
		-- or something trying to emulate a directory, probably.
		if (object and object ~= self and type(object) == "table" and object.FullyLoad) then
			object:FullyLoad()
		end
	end
end

--[[
	any? load_file(string path, any? base)
		path: The module path of the file.
		base: The root to pass to the module, defaults to N.Base.

	Loads a file and executes it, returning the result.
	Uses the built-in filesystem abstractions.
]]
local function load_file(file, base)
	local method = assert(load_with_env(file:Read(), file.Path))
	local result = method(base or N.Base, file.Path)

	return result
end

--[[
	Directory? load_directory(string path)
		path: The module path of the file.

	Loads a directory and its init file, returning the result.
	Uses the built-in filesystem abstractions.
]]
local function load_directory(directory)
	local object = N:Get(module_join(directory.Path, "_"))
	local initializing = {}

	object = dictionary_shallow_copy(directory_interface, object)

	object._directory = object._directory or directory

	setmetatable(object, {
		__index = function(self, key)
			local path = module_join(directory.Path, key)

			if (initializing[key]) then
				error(("Circular reference loading %q!"):format(path), 2)
			end

			initializing[key] = true

			local result = N:Get(path)
			self[key] = result

			initializing[key] = false

			return result
		end
	})

	return object
end

--===============--
-- NAMESPACE API --
--===============--

-- This library can be accessed by any codefile by using
-- Directory:GetNamespaceCore()
-- on any Namespace directory.

--[[
	void N:AddRebase(string pattern, any target)
		pattern: The pattern a path should match to use this rule.
		target: The object to be used as the root namespace upon success.

	Adds a rebasing rule for modules that match this rule.
	Used primarily for embedding existing namespace modules.

	Can still have issues if the called module interacts with the namespace core.
]]
function N:AddRebase(pattern, target)
	assert(type(pattern) == "string", "Bad argument #1 to N:AddRebase, must be a string!")

	table.insert(self._rebasing, {pattern, target})
end

--[[
	void N:ClearRebases()

	Removes all rebasing rules from the core.
]]
function N:ClearRebases()
	for key, value in pairs(self._rebasing) do
		self._rebasing[key] = nil
	end
end

--[[
	any? N:Get(string path)
		path: The path to the module, period delimitted

	Returns the object relative to this namespace's root, if it exists.
]]
function N:Get(path)
	path = path or ""

	-- Check for already loaded module!
	if (self._loaded[path]) then
		return self._loaded[path]
	end

	-- Run path through our rebasing rules
	local base = N.Base
	for i, rebase in ipairs(self._rebasing) do
		if (path:match(rebase[1])) then
			base = rebase[2] or base
		end
	end

	-- Is this a file?
	local file = N.FS:GetFile(path)

	if (file) then
		local object = load_file(file, base)
		file:Close()

		if (object) then
			self._loaded[path] = object
			return object
		end
	else
		-- How about a directory?
		local directory = N.FS:GetDirectory(path)

		if (directory) then
			local object = load_directory(directory)
			directory:Close()

			if (object) then
				self._loaded[path] = object
				return object
			end
		else
			return nil
		end
	end
end

-- If the Lib switch is set, make our base the core instead of the namespace itself.
if (N.Config.Lib) then
	N.Base = N:Get()
else
	N.Base = N
end

return N.Base